Use TP Tensor.metadata to carry device info

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44418 [WIP] Use streams from pool on RPC callees
* **#44396 Use TP Tensor.metadata to carry device info**
* #44417 Add missing rpc.shutdown()

Differential Revision: [D23602576](https://our.internmc.facebook.com/intern/diff/D23602576)

mrshenli committed:
Use TP Tensor.metadata to carry device info

[ghstack-poisoned]

mrshenli committed:
Update on "Use TP Tensor.metadata to carry device info"




[ghstack-poisoned]

mrshenli committed:
Update on "Use TP Tensor.metadata to carry device info"




[ghstack-poisoned]

mrshenli committed:
Update on "Use TP Tensor.metadata to carry device info"




[ghstack-poisoned]

mrshenli committed:
Update on "Use TP Tensor.metadata to carry device info"


Differential Revision: [D23602576](https://our.internmc.facebook.com/intern/diff/D23602576)

[ghstack-poisoned]

lw commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -538,26 +528,30 @@ std::tuple<tensorpipe::Message, TensorpipeWriteBuffers> tensorpipeSerialize(
       buffers.pickle.data(), buffers.pickle.size()});
-  for (const auto& tensor : pickler.tensorData()) {
-    const auto& tensorData = jit::getWriteableTensorData(tensor);
+  for (size_t i = 0; i < pickler.tensorData().size(); ++i) {
+    const auto& tensorData =
+        jit::getWriteableTensorData(pickler.tensorData()[i]);
```
Is `pickler.tensorData()` an efficient function that we can call repeatedly at no cost? Or should we store its value (or a reference to it)?

mrshenli replied:
Yep, it is an efficient function. Let me create a reference anyway incase it changes in the future.

https://github.com/pytorch/pytorch/blob/af9cad761ad1dd215470ebce58932b82d6ba8bee/torch/csrc/jit/serialization/pickler.h#L145-L147

lw commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -651,20 +634,21 @@ Message tensorpipeDeserialize(
+  for (size_t i = 0; i < message.tensors.size(); ++i) {
+    auto& tensor = message.tensors[i];
+    if (!tensor.metadata.empty()) {
+      TORCH_INTERNAL_ASSERT(
```
Is there a reason to have this check inside the for loop? It doesn't seem to depend on the value of `i`, does it?

mrshenli replied:
This check actually won't hold if tensors in RPC args or return values share storage. The pickler might pickle the storage instead the real tensors, and as a result, the tensor number in TpMessage and RpcMessage can be different. So I put a temporary check here to only verify the equivalence when there are non-empty metadata (i.e., newly introduced tests).  Will remove this when we can pass CUDA DataPtr to TP.

mrshenli committed:
Update on "Use TP Tensor.metadata to carry device info"


Differential Revision: [D23602576](https://our.internmc.facebook.com/intern/diff/D23602576)

[ghstack-poisoned]

mrshenli committed:
Update on "Use TP Tensor.metadata to carry device info"


Differential Revision: [D23602576](https://our.internmc.facebook.com/intern/diff/D23602576)

[ghstack-poisoned]
